src/puppetlabs/puppetdb/utils.clj

```diff
@@ -264,6 +264,55 @@
       (apply assoc m new-kvs)
       m)))
 
+(defn str-last
+  [s]
+  (if (= s "")
+    nil
+    (some-> s (.charAt (dec (.length s))))))
+
+(defn maybe-strip-escaped-quotes
+  [s]
+  (if (and (> (count s) 1) (= \" (first s) (str-last s)))
+    (subs s 1 (dec (count s)))
+    s))
+
+(defn parse-matchfields
+  [s]
+  (string/replace s #"match\((\".*\")\)" "$1"))
+
+(defn parse-indexing
+  [s]
+  (string/replace s #"\[(\d+)\]" ".$1"))
+
+(defn split-indexing
+  [path]
+  (flatten
+    (for [s path]
+      (if (re-find #"\[\d+\]$" s)
+        (-> s
+            (string/split #"(?=\[\d+\]$)")
+            (update 1 #(Integer/parseInt (subs % 1 (dec (count %))))))
+        s))))
+
+(defn handle-quoted-thing
+  [v]
+  (loop [result []
+         [s & splits] v]
+    (if (and (= \" (str-last s)) (not (= (count s) 1)))
+      [(string/join "." (conj result s)) splits]
+      (recur (conj result s) splits))))
+
+(defn dotted-query->fact-path
```

Comment:
I think I understand what this is doing. You have a dotted query string that may or may not have segments that are quoted. You want to split them by `.`, but if it's quoted, you want the quoting to denote the segment in case there is a `.` inside the quotes that you'd want to ignore.

I think I've got another idea for implementation, but could you put some tests together for these two functions, seems pretty easy to test and would be good to get some coverage there.